Adding buffered boolean param to FileSink and RollingFileSink

Pvlerick opened the pull request:
This is a rough draft implementation to address #650. I open this PR so we have code to discuss; this code is definitely not ready :wink: 

In all honesty, I don't like the boolean parameter too much. Very often boolean parameters are passed around to mean "behave in this way if true, in that way if false"; I think if two behaviours are distinct enough they should have they own classes (```BufferedFileSink```?).

The second point is the task that is started to process what is put in the queue. I don't like it too much either and I'm wondering if there would be a way to use the explicit implementation of ```IProducerConsumerCollection<T>```. I didn't look into this too much so I will explore, but maybe someone already has an idea.

Let me know what you think.

nblumhardt commented on src/Serilog.Sinks.File/Sinks/File/FileSink.cs:
```diff
@@ -88,17 +106,26 @@ static void TryCreateDirectory(string path)
             lock (_syncRoot)
             {
                 _textFormatter.Format(logEvent, _output);
                 _output.Flush();
```
Here is the `Flush()` call that `buffered: true` should switch off.

nblumhardt commented:
Hi Phillipe! Thanks for the PR.

The buffering I intended to refer to in the issue is the buffering provided under the hood by `System.IO.Stream` implementations.

> Both File and RollingFile flush the file stream after each write.

Turning "on" buffering would mean skipping the `Flush()` calls that these sinks make.

Sorry about the confusion, hope this clarifies it!

Pvlerick commented:
Oh I see now... Much simple than what I thought :smile: 

Ok then, I'll modify the PR accordingly.

Pvlerick commented:
@nblumhardt I reworked the approach a little bit. I added a ```autoFlush``` paramter to both ```FileSink``` and ```RollingFileSink``` that is relayed to the underlying ```StreamWriter```. With this, the call to ```Flush()``` can be suppressed and there is no need for an ```if``` statement.

Of course this parameter could be renamed to ```buffered``` and invert it when creating the ```StreamWriter```.

nblumhardt commented:
Thanks @Pvlerick!

A couple of thoughts -

 * `AutoFlush` will cause flushing after every write, which will be a lot more flushes than currently required; when events are formatted out as text there are a lot of fine-grained writes for each property etc. so the single `Flush()` call effectively batches these - using `AutoFlush` will most likely be a performance regression
 * Serilog by convention uses `false` as the default for all `Boolean` parameters, so `buffered = false` is preferred over `autoFlush = true` - sticking with this makes the API a bit more predictable

I think the `if()` version is probably the one to go with. All makes sense?

Cheers!

Pvlerick commented:
@nblumhardt makes sense. I updated the PR.

Pvlerick commented:
One more thing: should ```FileLoggerConfigurationExtensions.cs``` and ```RollingFileLoggerConfigurationExtensions.cs``` also be updated to take this into account, or is this left for advanced user that will not use the extension methods?

nblumhardt commented:
Looks great :+1:

I think this is a common enough scenario we should update the extension methods.

Pvlerick commented:
@nblumhardt all done.

nblumhardt commented:
Looking great! Nicely done.
